programs/monaco_protocol/src/instructions/market_liquidities/update_market_liquidities_with_cross_liquidity.rs

```diff
@@ -1,46 +1,153 @@
 use crate::error::CoreError;
-use crate::instructions::calculate_price_cross;
+use crate::state::market_account::Market;
 use crate::state::market_liquidities::{LiquiditySource, MarketLiquidities};
 use anchor_lang::prelude::*;
 
 pub fn update_market_liquidities_with_cross_liquidity(
+    market: &Market,
     market_liquidities: &mut MarketLiquidities,
     source_for_outcome: bool,
     source_liquidities: Vec<LiquiditySource>,
-    cross_liquidity: LiquiditySource,
 ) -> Result<()> {
     require!(
         market_liquidities.enable_cross_matching,
-        CoreError::MarketLiquiditiesUpdateError,
+        CoreError::MarketLiquiditiesCrossMatchingDisabled,
     );
 
-    // calculate price based on provided inputs
-    let source_prices = source_liquidities
-        .iter()
-        .map(|source_liquidity| source_liquidity.price)
-        .collect::<Vec<f64>>();
-
-    if let Some(cross_price) = calculate_price_cross(&source_prices) {
-        // provided cross_liquidity.price is valid
-        if cross_price == cross_liquidity.price {
-            if source_for_outcome {
-                market_liquidities.update_cross_liquidity_against(&source_liquidities);
-            } else {
-                market_liquidities.update_cross_liquidity_for(&source_liquidities);
-            };
-        }
-    }
+    require!(
+        source_liquidities.len() == (market.market_outcomes_count - 1) as usize,
+        CoreError::MarketLiquiditiesSourceLiquiditiesInvalid,
+    );
+
+    // ensuring that all newly created cross liquidity has sources sorted the same way
+    let mut source_liquidities_sorted = source_liquidities.to_vec();
+    source_liquidities_sorted.sort_by(|a, b| a.outcome.partial_cmp(&b.outcome).unwrap());
+    source_liquidities_sorted.dedup_by(|a, b| a.outcome == b.outcome);
+
+    // making sure there were no duplicates
+    require!(
+        source_liquidities_sorted.len() == source_liquidities.len(),
+        CoreError::MarketLiquiditiesSourceLiquiditiesInvalid,
+    );
```

Comment:
It might also be nice to filter out any sources with outcome that is `>= market.market_outcomes_count`.

Such a case will fail later but with a panic rather than a controlled error.

Reply:
Good point.